### PR TITLE
IT-5108 | add ProgressDonutChart

### PR DESCRIPTION
**Pull request by tweetgeek**

### CODE REVIEW

File: lib/js/components/ProgressDonutChart/ProgressDonutChart.vue [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+		},
+		isDone: { type: Boolean, default: false },
+		hasOverage: { type: Boolean, default: false },
+		ranges: {
+			type: Array as PropType<Array<ProgressDonutChartRange>>,
+			required: true,
+			validator(ranges) {
+				return ranges.every(
+					(range: ProgressDonutChartRange) =>
+						range.start >= 0 && range.length >= 0 && range.start + range.length <= 100,
```
Comment (kvas-damian) [inline lib/js/components/ProgressDonutChart/ProgressDonutChart.vue:262]: when I pass `length=0` then entire circle is rendered

Comment (kvas-damian) [inline lib/js/components/ProgressDonutChart/ProgressDonutChart.vue:262]: start and length doesn't work correctly when I pass floating number

File: lib/js/components/ProgressDonutChart/ProgressDonutChart.vue [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+		transform-origin: 50% 50%;
+	}
+
+	&__thumb {
+		stroke: $color-neutral-data-ghost;
+	}
+
+	&__track {
+		@for $i from 1 through 100 {
+			&[data-value='#{$i}'] {
```
Comment (kvas-damian) [inline lib/js/components/ProgressDonutChart/ProgressDonutChart.vue:142]: wow, it'll generate 99 css blocks (btw. 0 is missing). It's a lot of CSS.

I see 2 alternatives. 
1. Use css variable and pass it through style prop: `style="--length:30"`
2. Pass entire `stroke-dashoffset` with value into style prop

File: lib/js/components/ProgressDonutChart/ProgressDonutChart.vue [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+			content: '';
+			display: inline-block;
+			position: absolute;
+			top: 0;
+		}
+
+		&::before {
+			left: -4px;
+			width: 2px;
+			height: 2px;
```
Comment (kvas-damian) [inline lib/js/components/ProgressDonutChart/ProgressDonutChart.vue:213]: should we use const for size of the dot?

File: lib/js/components/ProgressDonutChart/ProgressDonutChart.vue [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+		label: { type: String, default: '' },
+		state: {
+			type: String,
+			default: PROGRESS_DONUT_CHART_STATES.DEFAULT,
+			validator(state) {
+				return Object.values(PROGRESS_DONUT_CHART_STATES).includes(state);
+			},
+		},
+		isDone: { type: Boolean, default: false },
+		hasOverage: { type: Boolean, default: false },
```
Comment (kvas-damian) [inline lib/js/components/ProgressDonutChart/ProgressDonutChart.vue:255]: @rogatty suggested to remove these props and use only `state`. Did you discuss it with UX?

File: lib/js/components/ProgressDonutChart/ProgressDonutChart.stories.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+			{
+				color: PROGRESS_DONUT_CHART_RANGE_COLORS.INFO,
+				start: 30,
+				length: 25,
+			} as ProgressDonutChartRange,
+			{
+				color: PROGRESS_DONUT_CHART_RANGE_COLORS.SUCCESS,
+				start: 0,
+				length: 30,
+			} as ProgressDonutChartRange,
```
Comment (kvas-damian) [inline lib/js/components/ProgressDonutChart/ProgressDonutChart.stories.ts:52]:
```suggestion
			{
				color: PROGRESS_DONUT_CHART_RANGE_COLORS.SUCCESS,
				start: 0,
				length: 30,
			} as ProgressDonutChartRange,
			{
				color: PROGRESS_DONUT_CHART_RANGE_COLORS.INFO,
				start: 30,
				length: 25,
			} as ProgressDonutChartRange,
			{
				color: PROGRESS_DONUT_CHART_RANGE_COLORS.FAIL_GHOST,
				start: 55,
				length: 25,
			} as ProgressDonutChartRange,
```

File: lib/js/components/ProgressDonutChart/ProgressDonutChart.vue [hunk 6]
```diff
@@ -0,0 +1,287 @@
+<template>
+	<div
+		:class="{
+			progressDonutChart: true,
+		}"
+	>
+		<svg
+			class="progressDonutChart__svg"
+			width="40"
```
Comment (kvas-damian) [inline lib/js/components/ProgressDonutChart/ProgressDonutChart.vue:9]: maybe we should have it as a const in JS?

File: lib/js/components/ProgressDonutChart/ProgressDonutChart.vue [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+@import '../../../styles/settings/spacings';
+@import '../../../styles/settings/typography/tokens';
+
+$progress-donut-chart-width: 40px;
+$progress-donut-chart-height: 40px;
+$progress-donut-chart-circle-stroke-width: 4;
+$pi: 3.1416;
+$progress-donut-chart-circle-radius: 18; // (width / 2) - (stroke-width / 2)
+$circle-circumference: 2 * $pi * $progress-donut-chart-circle-radius;
+$progress-donut-chart-loading-dot-width: 2px;
```
Comment (kvas-damian) [inline lib/js/components/ProgressDonutChart/ProgressDonutChart.vue:73]: you could use one variable: `$progress-donut-chart-loading-dot-size`

File: lib/js/components/ProgressDonutChart/ProgressDonutChart.vue [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+
+	&__loaderText {
+		position: relative;
+		width: $progress-donut-chart-loading-dot-width;
+		height: $progress-donut-chart-loading-dot-height;
+		margin-top: $space-xxxs;
+		border-radius: 100%;
+		background-color: $color-neutral-text;
+		color: $color-neutral-text;
+		animation: dot-flashing 1s infinite linear alternate;
```
Comment (kvas-damian) [inline lib/js/components/ProgressDonutChart/ProgressDonutChart.vue:201]: IMO, this animation should be faster

File: lib/js/components/ProgressDonutChart/ProgressDonutChart.vue [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+}
+
+.progressDonutChart {
+	position: relative;
+	width: $progress-donut-chart-width;
+	height: $progress-donut-chart-height;
+	padding: 0;
+	margin: 0;
+	overflow: hidden;
+	box-sizing: border-box;
```
Comment (kvas-damian) [inline lib/js/components/ProgressDonutChart/ProgressDonutChart.vue:121]:
```suggestion
	overflow: hidden;
```

Why do we need this attributes?

File: lib/js/components/ProgressDonutChart/ProgressDonutChart.vue [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+
+	&__svg {
+		display: block;
```
Comment (kvas-damian) [inline lib/js/components/ProgressDonutChart/ProgressDonutChart.vue:124]: is it required?

File: lib/js/components/ProgressDonutChart/ProgressDonutChart.vue [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+
+		display: flex;
+		position: absolute;
+		top: 0;
+		left: 0;
+		width: 100%;
+		height: 100%;
+		justify-content: center;
+		flex-wrap: nowrap;
+		align-items: center;
```
Comment (kvas-damian) [inline lib/js/components/ProgressDonutChart/ProgressDonutChart.vue:176]: alphabetical order?

File: lib/js/components/ProgressDonutChart/ProgressDonutChart.vue [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+		@include info-m-default-bold;
+
+		display: flex;
+		position: absolute;
+		top: 0;
+		left: 0;
+		width: 100%;
+		height: 100%;
+		justify-content: center;
+		flex-wrap: nowrap;
```
Comment (kvas-damian) [inline lib/js/components/ProgressDonutChart/ProgressDonutChart.vue:175]: isn't it by default?

File: lib/js/components/ProgressDonutChart/ProgressDonutChart.vue (hunk 11 above)
Comment (rogatty) [reply lib/js/components/ProgressDonutChart/ProgressDonutChart.vue:176]: @kvas-damian enable https://github.com/hudochenkov/stylelint-order/blob/master/rules/properties-alphabetical-order/README.md if you think it's valuable 🙂

File: lib/js/components/ProgressDonutChart/ProgressDonutChart.vue [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+		}
+
+		&::after {
+			left: $progress-donut-chart-loading-dot-width * 2;
+			width: $progress-donut-chart-loading-dot-width;
+			height: $progress-donut-chart-loading-dot-height;
+			border-radius: 100%;
+			background-color: $color-neutral-text;
+			color: $color-neutral-text;
+			animation: dot-flashing 1s infinite alternate;
```
Comment (kvas-damian) [inline lib/js/components/ProgressDonutChart/ProgressDonutChart.vue:230]: cound be moved to `&::before, &::after`

File: lib/js/components/ProgressDonutChart/ProgressDonutChart.vue [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+			/>
+			<template v-else>
+				<circle
+					v-for="(range, index) in calculatedRanges"
+					:key="`circle_${index}`"
+					class="progressDonutChart__circle progressDonutChart__track"
+					cx="20"
+					cy="20"
+					r="18"
+					:style="`--length: ${range.length}; transform: rotate(calc(90deg + ${range.rotate}deg));`"
```
Comment (kvas-damian) [inline lib/js/components/ProgressDonutChart/ProgressDonutChart.vue:34]: should we add a unit test?

File: lib/js/components/ProgressDonutChart/ProgressDonutChart.vue (hunk 11 above)
Comment (kvas-damian) [reply lib/js/components/ProgressDonutChart/ProgressDonutChart.vue:176]: stop spying us. You have a lot to do in other ticket :trollface:

Comment (tweetgeek) [reply lib/js/components/ProgressDonutChart/ProgressDonutChart.vue:176]: maybe we should enable https://github.com/hudochenkov/stylelint-order/blob/master/rules/properties-alphabetical-order/README.md ?

Comment (kvas-damian) [reply lib/js/components/ProgressDonutChart/ProgressDonutChart.vue:176]: @rogatty https://github.com/bethinkpl/design-system/pull/276

File: lib/js/components/ProgressDonutChart/ProgressDonutChart.vue [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+	},
+	computed: {
+		labelText() {
+			return [this.state === PROGRESS_DONUT_CHART_STATES.OVERAGE && '+', this.label]
+				.filter(Boolean)
+				.join('');
+		},
+		calculatedRanges() {
+			return this.ranges.map((range) => {
+				return { ...range, rotate: (range.start / 100) * 360 + 90 };
```
Comment (kvas-damian) [inline lib/js/components/ProgressDonutChart/ProgressDonutChart.vue:283]:
```suggestion
				// `+ 90` is to set starting point at the bottom
				return { ...range, rotate: (range.start / 100) * 360 + 90 };
```

**Review by kvas-damian — APPROVED** [review]

File: lib/js/components/ProgressDonutChart/ProgressDonutChart.vue (hunk 14 above)
Comment (kvas-damian) [reply lib/js/components/ProgressDonutChart/ProgressDonutChart.vue]: [`72ec278` (#274)](https://github.com/bethinkpl/design-system/pull/274/commits/72ec278a74c1d268b6cbb5fa8606f72dcfa0b9c3)

File: lib/js/components/ProgressDonutChart/ProgressDonutChart.vue [hunk 16]
```diff
@@ -0,0 +1,292 @@
...
+		height: 100%;
+		justify-content: center;
+		left: 0;
+		position: absolute;
+		top: 0;
+		width: 100%;
+	}
+
+	&__icon {
+		color: $color-success-text;
```
Comment (kvas-damian) [inline lib/js/components/ProgressDonutChart/ProgressDonutChart.vue:176]:
```suggestion
		color: $color-success-icon;
```

File: lib/js/components/ProgressDonutChart/ProgressDonutChart.vue (hunk 8 above)
Comment (kvas-damian) [reply lib/js/components/ProgressDonutChart/ProgressDonutChart.vue]: [`72ec278` (#274)](https://github.com/bethinkpl/design-system/pull/274/commits/72ec278a74c1d268b6cbb5fa8606f72dcfa0b9c3)

**Pull request merged**